[WIP] fix volume control for audio sources

antis81 opened the pull request:
The next milestone forward to support reliable audio calls. Anyone interested in improving audio calls, please let me know and help review/test.
- [x] Clean up audio settings ui.
- [x] Determine current master volume from qTox settings in output device initialization.
- [x] ~~Bind gain/volume controls to system settings.~~ OpenAL cannot control system volume and it's not the right thing to do anyway.
- [x] Remove destructive input gain algorithm:
  - Calculation of input gain is plain wrong, leading to destroyed input values.
- [ ] Replace the gain algorithm with something more suitable.
- [x] Remove invalid self-creation of unregistered audio source in `playGroupAudio`.
- [x] Moved to #2785 -> ~~Return false when initializing device with specifier "none".~~
- [x] Fix initialization of A/V settings:
  - [x] Do not open the audio device in constructor (called, when qTox starts)
  - [x] Fix initialization of in/out volumes (audio device must be initalized and valid **before**)
- [x] Add "test output" button to A/V settings form. This is how it looks like:
  - ![play-test-sound](https://cloud.githubusercontent.com/assets/440517/12374862/8cb5a2d6-bcaa-11e5-9e29-0c957e7081a4.png)
- [x] Disable volume sliders, when no device is selected.

**KNOWN BUGS THAT WON'T BE FIXED HERE:**
- :beetle: :+1: The introduced "play test sound" button adds a nice penetration check for `AudioPlayer`, that reveals an unrelated bug, when playing the same sound ~100 times.

antis81 committed:
initialize a subscribed audio source with "master" volume instead of "max."

antis81 committed:
clean up audio settings ui

antis81 commented:
Cleaned up audio settings ui:
    - removed min/max labels, which don't have any meaning
    - changed text of "volume" and "gain" labels (maybe "input gain" and "output volume"?)
        --> German translation: `volume ==> Lautstärke` and `gain ==> Eingabeempfindlichkeit`.
    ![audio-settings](https://cloud.githubusercontent.com/assets/440517/12078690/eb0ba37e-b21b-11e5-89dd-b8f178e0078c.png)

antis81 commented:
@ovalseven8 Could you try your headset setup with this PR? You should get better results with this and the strange sounds might go away even in more complex setups. However to finish this, there's a whole bunch of concept work needed - thus I'll do this step by step.

antis81 commented:
@agilob @r-ku @towlie Could you please review?

Works for me in the means of improving overall audio volume experience. Output is a little low, but the unwanted echo effects definitely disappear. Thus, when raising the system environment's gain & volume, I get a very clean result, without any unwanted noise.

towlie commented:
Will test tomorrow.

towlie commented:
Audio calls don't work. I can hear nothing. But the audio meter shows something on both sides.
https://gist.github.com/towlie/c27eaddc52c3aad92971
And why do you use 100 now instead of 400 for mic gain? Well I couln't test that change because you can hear nothing ^^.

antis81 commented:
@towlie You should now hear something again. :) Sorry, I cannot give any explanations right now, because I'm still cleaning up some **really** crappy code from previous implementation but it's getting more and more clear. Anyway FYI, input gain currently doesn't have any effect (the brandnew input source is invalid) - still need to figure that out. Thus, going a step back marking it WIP again.

towlie commented:
It's still not working. https://gist.github.com/towlie/a2081f886bd8002980d3

antis81 commented:
@towlie Could you do another check please? We now have a real master volume control. Input gain still not available. This will require more investigation.

towlie commented:
- output device not stopped/paused on start (only if a sound was played)
- output device not stopped when leaving the av tab (only if a sound was played)
- audio meter sometimes doesn't work (you have to retry and hope for luck 50:50 chance)

these 2 are related (if one doesn't occur then the other happens): 
- audio call is sometimes borked when the audio/video tab is open while in call (disappears if you leave the tab)
- sometimes the audio meter doesn't work while in call

https://gist.github.com/towlie/cc770d12ef92c6024070

towlie commented:
```
0 : Warning: QWaitCondition: Destroyed while threads are still waiting
```

This wasn't there before.

antis81 commented:
Btw.: Which OS are you checking on?

> output device not stopped/paused on start (only if a sound was played)

What do you mean by that?

> This wasn't there before.

It was there before (check against master; issue ~~#2703~~, #2705). This origins from the `AudioMeterListener`, which is still not implemented correctly. It actually does a (fast) asynchronous read on input and this is, why it's accuracy varies. I'll fix this in another PR.

Thanks for checking. This confirms, that the "master volume" implementation works and I can now focus on the input gain.

towlie commented:
@antis81 
It means that I started qTox and I see that an output device is open (Higher CPU usage in the audioservice), which is not needed and can be closed by canceling calls or getting message beeps.
(Windows)

antis81 commented:
@towlie This is off-topic to this PR and I cannot reproduce on Linux -> ALSA.

Anyway, an audio device gets initialized through one of these methods **only**:
1. `Audio::subscribeInput()`
2. `Audio::subscribeOutput()`
3. `Audio::reinitInput()`
4. `Audio::reinitOutput()`

Each of these methods leaves an entry in the log file, so you will see it.

towlie commented:
I think it is not because it doesn't happen on master.

towlie commented:
I can clearly see that you create two handles on start:
ALPC Port   \BaseNamedObjects\AudioEngineDuplicateHandleApiPort964541503
Event   \BaseNamedObjects\AudioEndPoint_PlaybackReady
Which doesn't happen on master.

---

~~Did you mix this with other ungmerged PR?~~

antis81 commented:
@towlie Your logfile should mention something like "opening audio input/output device" when starting qTox. Surprisingly we don't have that issue on linux. Anyway, I'm going to rewrite some UI stuff to make it really good :)

antis81 committed:
major changes to audio volume control

* use the ALListener object for master volume control
* initialize audio sources
* audio volumes are now expressed as percentage values between 0 and 1 to the public API
* removed artificial amplification of input samples
* removed invalid audio source generation in group calls
* minor: fixed gain slider max. value to 100 percent

antis81 committed:
initialize volume sliders in av-settings ui at the right place

antis81 committed:
use utf-8 strings for audio device specifiers

antis81 commented:
@towlie Problem solved :)

towlie commented:
@antis81 Which one?
I guess you are talking about this open output device thing. I don't see it happen anymore.
I wouldn't wonder if this is just coincidence. ;)

towlie commented:
I didn't see anything broken but the log spits out an error:
https://gist.github.com/towlie/aac9840875168047c174
I cannot tell you what this is.

```
[02:33:43.144] ../src/audio/audio.cpp:765 : Warning: OpenAL error: 40961
```

antis81 commented:
> I guess you are talking about this open output device thing. I don't see it happen anymore.
> I wouldn't wonder if this is just coincidence. ;)

No coincidence :). It is exactly caused by wrong initialization of the volume sliders in A/V settings form (this is invisible to users). I'll move code into the right place.

antis81 committed:
fix getter for audio master output volume

Return the ALListener's gain as "master" volume.

antis81 committed:
remove unneeded "setCurrentIndex(-1)", when reading audio in/out device lists

The combo boxes current index is set to the correct value at end of initialization loop.

antis81 committed:
initialize gain/volume after initialization of audio in/out device

The gain/volume values are now applied to the settings once, when releasing the slider instead of continuously.

antis81 committed:
change method description of Audio::setOutputVolume to what it should do

antis81 committed:
add "play test sound" button, while changing volume

antis81 committed:
enable volume sliders only, if a valid audio device is selected

antis81 committed:
do not warn, when requesting the volume from an invalid audio device

Instead, we return a defined value (0.0) and keep quiet otherwise.

antis81 commented:
@towlie It's getting better… :). I'd be happy, if you could do another check and provide feedback from a user's view point.

Improvements:
- There are no more "dangling" audio devices after closing/leaving the A/V settings tab.
- Volumes are now correctly initialized to the current values (gain is always 0.0 -> see below).
- When selecting "No Audio Device", the volume sliders are now correctly set to 0.0 and disabled.

Known issues:
- The new introduced "play test sound" button adds a nice penetration check and triggers an unrelated bug in `AudioPlayer` -> it works (with warnings), when a call is active.
- The VU meter (the coloured widget) is unstable.
- Mic. gain control is still not available.

agilob commented:
"Play test sound" doesn't work for me at all.

antis81 commented:
@agilob You mean it doesn't play a sound when moving the slider?

agilob commented:
It doesn't play anything, ever:
Sorry, forgot about log.

```
[11:49:55.405] src/persistence/profile.cpp:314 : Debug: Saving tox save to  "/home/agilob/.config/tox/agilob.tox"
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
[11:50:19.274] src/audio/audio.cpp:489 : Debug: Opening audio output "Built-in Audio Analog Stereo"
[11:50:19.277] src/audio/audio.cpp:514 : Debug: Opened audio output "Built-in Audio Analog Stereo"
[11:50:19.288] src/audio/audio.cpp:445 : Debug: Opening audio input "Built-in Audio Analog Stereo"
[11:50:19.290] src/audio/audio.cpp:475 : Debug: Opened audio input "Built-in Audio Analog Stereo"
[11:50:19.350] src/audio/audio.cpp:402 : Debug: Subscribed to audio input device [ 1 subscriptions ]
[11:50:20.464] src/audio/audio.cpp:350 : Warning: OpenAL error: 40964
…
[11:50:22.784] src/audio/audio.cpp:350 : Warning: OpenAL error: 40964
[11:50:22.830] src/audio/audio.cpp:701 : Debug: Closing audio output
```

antis81 commented:
@agilob Please read description or comment above. :) To actually hear something, system volume has to be "loud enough". Simple as that.

tux3 commented:
Getting an OpenAL error 40964 ("Invalid operation") sounds like a qTox bug.

antis81 commented:
Well, it is. And @tux3 I wrote yesterday on chat, that this happens, plus I wrote it 2 times above in description and in the comments, that this is a **known bug**. AudioPlayer is simply not made for running ~50 threads, playing all the same buffer in parallel. Isn't that a nice penetration test for AudioPlayer? :)

Anyway I don't actually care as this is completely off-topic. If someone likes to improve the AudioPlayer, feel free to code it.

tux3 commented:
@antis81 Sorry, I haven't reviewed the code yet. I'm not sure I understand, why 50 threads? What are the threads doing?
I understand that this is a bit off-topic, but if it's a bug that prevents testing other things, well...

towlie commented:
@antis81 There are still "dangling" audio devices after closing the av tab.
https://gist.github.com/towlie/e076b966e255cdfb5b14

antis81 commented:
@tux3 

> Sorry, I haven't reviewed the code yet.

No problem. I'm hoping for (better) ideas and actual improvements to the audio related code base.

> I'm not sure I understand, why 50 threads?

Uhm, because currently each call to `Audio::playMono16Sound()` will start a threaded `AudioPlayer`. This works fine under "normal" conditions (like e.g. an incoming call notification during an active call), but not when playing sound files in a fast sequence like when moving the volume slider.

> I understand that this is a bit off-topic, but if it's a bug that prevents testing other things, well...

Which tests does it prevent? It only makes it a bit easier to actually hear, that an output volume of 100% actually matches the system's output volume (which is our maximum). Thus, we know we're on the right track, but not yet reached the finish line. Improving `AudioPlayer` is another PR (it will result in a rewrite; I checked that today).

tux3 commented:
I'm going to pull this on a local `audio-next` branch so it can be worked on further.

tux3 commented on src/audio/audio.cpp:
```diff
@@ -56,6 +56,22 @@
+    if (alc_err) { qWarning("OpenAL error: %d", alc_err); } \
+    } while (0)
+#endif
+
```
Those should be static inline functions, not macros.

antis81 replied:
Ok, going to change that later today.

tux3 replied:
Don't worry, I'll be working on it

tux3 commented:
So, as I understand it, the input gain in your PR is completely unused, right?
You removed the old algorithm but you didn't replace it with anything

antis81 commented:
Yes. Because this "anything" does not exist yet. If you have an idea how to realize that I'd be happy.

tux3 commented:
Wouldn't multiplying every captured sample by the gain be enough?
That should even be a fast SIMD-able operation since we don't need to clamp results (because 0<=gain<=1)

tux3 commented:
I cleanup up the audio input part a little, feel free to look at the audio-next branch.

I still need to look over the audio output part and fix any bugs.

TheNain38 commented:
Can be closed as it was merged in the audio-next branch? Or it could be closed when the gain algorithm is replaced?
Also, something doesn't compile on https://build.tox.chat/job/qTox_build_windows_x86-64_release/475/console

antis81 commented:
Will close this as some concepts diverged.

tux3 commented:
I still need to work some more on the audio branch because I think it's not stable yet, but I merged the what we have so far to fix the biggest issues.
@antis81 Feel free to send PRs to the audio branch if you want :)

antis81 commented:
That's totally fine for me. I'll experiment with the gain control further and I have something in mind that might actually work. Though the idea of tweaking the input data doesn't seem to be completely wrong direction. Btw. the filter needs to be redone as well.
